File: desktop/src/test/java/net/sf/freecol/common/model/ColonyTest.java
package net.sf.freecol.common.model;

import static org.junit.jupiter.api.Assertions.*;

import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.backends.lwjgl3.Lwjgl3Files;
import com.badlogic.gdx.utils.ObjectIntMap;

import net.sf.freecol.common.model.Colony.NoBuildReason;
import net.sf.freecol.common.model.player.Player;
import net.sf.freecol.common.model.specification.BuildingType;
import promitech.colonization.savegame.SaveGameParser;

public class ColonyTest {

	Game game;
	Player dutch;
	BuildingType stockadeType;
	
    @BeforeAll
    public static void beforeClass() {
        Gdx.files = new Lwjgl3Files();
    }
	
    @BeforeEach
    public void setup() throws Exception {
    	game = SaveGameParser.loadGameFormClassPath("maps/savegame_1600_for_jtests.xml");
    	dutch = game.players.getById("player:1");
		stockadeType = Specification.instance.buildingTypes.getById("model.building.stockade");
    }
    
	@Test
	public void shouldDisallowBuildUnit() throws Exception {
		// given
        Colony colony = dutch.settlements.getById("colony:6528").asColony();
        
        UnitType freeColonist = Specification.instance.unitTypes.getById("model.unit.freeColonist");
        
		// when
        NoBuildReason noBuildReason = colony.getNoBuildReason(freeColonist);
		
		// then
        assertEquals(NoBuildReason.MISSING_BUILD_ABILITY, noBuildReason);
	}

	@Test
	public void canCalculateTurnsToCompleteBuilding() throws Exception {
		// given
		Colony colony = dutch.settlements.getById("colony:6528").asColony();
		colony.getGoodsContainer().increaseGoodsQuantity("model.goods.hammers", 100);
		colony.getGoodsContainer().decreaseToZero("model.goods.tools");

		BuildingType runDistilleryType = Specification.instance.buildingTypes.getById("model.building.rumDistillery");
		
		// when
		ObjectIntMap<String> requiredTurnsForGoods = new ObjectIntMap<String>(2);
		int turnsToComplete = colony.getTurnsToComplete(runDistilleryType, requiredTurnsForGoods);

		// then
		assertEquals(Colony.NEVER_COMPLETE_BUILD, turnsToComplete);
	}

}

File: desktop/src/test/java/net/sf/freecol/common/model/ColonyTest.kt
package net.sf.freecol.common.model

import net.sf.freecol.common.model.Colony.NoBuildReason
import net.sf.freecol.common.model.specification.GoodsType
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import promitech.colonization.savegame.Savegame1600BaseClass

class ColonyTest : Savegame1600BaseClass() {

    @Test
    fun shouldDisallowBuildUnit() {
        // given
        val freeColonist = Specification.instance.unitTypes.getById(UnitType.FREE_COLONIST)

        // when
        val noBuildReason = nieuwAmsterdam.getNoBuildReason(freeColonist)

        // then
        assertThat(noBuildReason).isEqualTo(NoBuildReason.MISSING_BUILD_ABILITY)
    }

    @Test
    fun canCalculateTurnsToCompleteBuilding() {
        // given
        val colony = nieuwAmsterdam
        colony.goodsContainer.increaseGoodsQuantity(GoodsType.HAMMERS, 100)
        colony.goodsContainer.decreaseToZero(GoodsType.TOOLS)
        val runDistilleryType = Specification.instance.buildingTypes.getById("model.building.rumDistillery")

        // when
        val buildProgress = BuildProgress.calculateBuildProgress(colony, runDistilleryType)

        // then
        assertThat(buildProgress.turnsToComplete).isEqualTo(BuildProgress.LACK_OF_RESOURCES)
        assertThat(buildProgress.isCompleted).isFalse
        assertThat(buildProgress.noComponentsInProduction()).isTrue
    }

    @Test
    fun should_calculate_finished_building() {
        // given
        val colony = nieuwAmsterdam
        colony.goodsContainer.increaseGoodsQuantity(GoodsType.HAMMERS, 100)
        colony.goodsContainer.increaseGoodsQuantity(GoodsType.TOOLS, 20)
        val runDistilleryType = Specification.instance.buildingTypes.getById("model.building.rumDistillery")

        // when
        val buildProgress = BuildProgress.calculateBuildProgress(colony, runDistilleryType)

        // then
        assertThat(buildProgress.turnsToComplete).isEqualTo(0)
        assertThat(buildProgress.isCompleted).isTrue
        assertThat(buildProgress.noComponentsInProduction()).isFalse
        assertThat(buildProgress.isProgressStopedBecauseOfLackOfComponents).isFalse
    }


}